RaftMongoWithRaftReconfig.tla

```diff
@@ -0,0 +1,276 @@
+                                      ELSE IF index = i THEN Leader ELSE Follower]
+    /\ currentTerm' = [index \in Server |-> IF \/ index \in ayeVoters 
+                                               \/ index = i
+                                            THEN currentTerm[i] + 1 
+                                            ELSE currentTerm[index]]
+    /\ LET noop == [term  |-> currentTerm[i] + 1, configVersion |-> GetConfigVersion(i)]
+           newLog == Append(log[i], noop)
+       IN  log' = [log EXCEPT ![i] = newLog]
+    /\ UNCHANGED <<commitPoint, configs>>
+
+\* ACTION
+\* Leader i receives a client request to add v to the log.
+ClientWrite(i) ==
+    /\ state[i] = Leader
+    /\ LET entry == [term  |-> currentTerm[i], configVersion |-> GetConfigVersion(i)]
+           newLog == Append(log[i], entry)
+       IN  log' = [log EXCEPT ![i] = newLog]
+    /\ UNCHANGED <<serverVars, configs>>
+        
+\* ACTION
+AdvanceCommitPoint ==
+    \E leader \in Server : \E acknowledgers \in SUBSET Server :
+        /\ state[leader] = Leader
+        /\ acknowledgers \subseteq Agree(leader, Len(log[leader]))
+        /\ acknowledgers \in Quorum(leader)
+        \* If we comment out the following line, a replicated log entry from old primary will voilate the safety.
+        \* [ P (2), S (), S ()]
+        \* [ S (2), S (), P (3)]
+        \* [ S (2), S (2), P (3)] !!! the log from term 2 shouldn't be considered as committed.
+        /\ LogTerm(leader, Len(log[leader])) = currentTerm[leader]
+        \* If an acknowledger has a higher term, the leader would step down.
+        /\ \A j \in acknowledgers : currentTerm[j] <= currentTerm[leader]
+        /\ commitPoint' = [commitPoint EXCEPT ![leader] = [term |-> LastTerm(log[leader]), index |-> Len(log[leader])]]
+        /\ UNCHANGED <<electionVars, logVars, configs>>
+        
+\* Return whether Node i can learn the commit point from Node j.
+CommitPointLessThan(i, j) ==
+   \/ commitPoint[i].term < commitPoint[j].term
+   \/ /\ commitPoint[i].term = commitPoint[j].term
+      /\ commitPoint[i].index < commitPoint[j].index
+
+\* ACTION
+\* Node i learns the commit point from j via heartbeat with term check
+LearnCommitPointWithTermCheck(i, j) ==
+    /\ CommitPointLessThan(i, j)
+    /\ LastTerm(log[i]) = commitPoint[j].term
+    /\ commitPoint' = [commitPoint EXCEPT ![i] = commitPoint[j]]
+    /\ UNCHANGED <<electionVars, logVars, configs>>
+       
+UpdateTermThroughHeartbeat(i, j) ==
+    /\ j \in ServerViewOn(i)  \* j is in the config of i.
+    /\ currentTerm[j] > currentTerm[i]
+    /\ currentTerm' = [currentTerm EXCEPT ![i] = currentTerm[j]]
+    /\ state' = [state EXCEPT ![i] = IF ~(state[i] = Leader) THEN state[i] ELSE Follower]
+    /\ UNCHANGED <<commitPoint, logVars, configs>>
+        
+Reconfig(i, newConfig) ==
+    /\ state[i] = Leader
+    /\ i \in newConfig
+    /\ Cardinality(ServerViewOn(i) \ newConfig) + Cardinality(newConfig \ ServerViewOn(i)) <= 1
```

Comment:
This is clever!

Reply:
Thank you! I'll add a comment here to clarify this represents single server addition/removal.